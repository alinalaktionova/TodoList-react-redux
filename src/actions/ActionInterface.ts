import {
  ADD_TODO,
  DELETE_ALL,
  REMOVE_TODO,
  TOGGLE_COMPLETE_TODO
} from "../constants/ActionTypes";
import {Dispatch} from "react";

export interface AddTodo {
  type: typeof ADD_TODO;
  id: number;
  text: string;
}
export interface ToggleTodoComplete {
  type: typeof TOGGLE_COMPLETE_TODO;
  id: number;
}
export interface RemoveTodo {
  type: typeof REMOVE_TODO;
  id: number;
}
export interface DeleteAll {
  type: typeof DELETE_ALL;
}

export type ActionType = AddTodo | ToggleTodoComplete | RemoveTodo | DeleteAll ;

