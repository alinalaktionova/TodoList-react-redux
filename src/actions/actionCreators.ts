import {ADD_TODO, TOGGLE_COMPLETE_TODO} from "../constants/ActionTypes";
import {REMOVE_TODO} from "../constants/ActionTypes";
import {DELETE_ALL} from "../constants/ActionTypes";
import {ActionType} from "./ActionInterface";


export let id = 0;

export const addTodo = (text: string) => ({
         type: ADD_TODO,
         id: id++,
         text
       });

export const toggleTodoComplete = (id: number) => ({
    type: TOGGLE_COMPLETE_TODO,
    id,
});

export const removeTodo = (id: number) => ({
    type: REMOVE_TODO,
    id
});

export const deleteAll = () => ({
    type: DELETE_ALL,
});




