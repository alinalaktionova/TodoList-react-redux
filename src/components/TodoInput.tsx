import React, {Dispatch} from "react";
import { connect } from "react-redux";
import {id} from "../actions/actionCreators";
import styled from "styled-components";
import {ActionType} from "../actions/ActionInterface";
import {ADD_TODO} from "../constants/ActionTypes";
import PropsInterface from "../interfaces/PropsInterface";

const TodoInputForm = styled.div`
  margin: 0 auto;
  width: 70%;
  display: flex;
  flex-direction: row;
`;

const InputField = styled.input`
  width: 85%;
  border-radius: 7px;
  overflow: auto;
  background-color: white;
  box-shadow: 0px 0.25rem 1rem rgba(black, 0.25);
`;

const StyledButton = styled.button`
  width: 15%;
  margin: 0 auto;
  padding: 5px;
  border-radius: 5px;
  background: dimgrey;
  color: white;
  font-size: 16px;
  text-transform: capitalize;
  letter-spacing: 1px;
  display: inline-block;
  cursor: pointer;
`;


let idEnc = id;

const TodoInput = (props: Partial<PropsInterface>) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  let [input, setInputState] = React.useState("");
  function handleInputChange(event: React.ChangeEvent<HTMLInputElement>) {
    setInputState(event.target.value);
  }

  return (
    <TodoInputForm>
      <InputField
        type="text"
        placeholder="add a todo item"
        onChange={event => handleInputChange(event)}
        ref={inputRef}
      />
      <StyledButton
        type="button"
        onClick={(event: React.MouseEvent) => {
          props.addTodo(input);
          if (inputRef && inputRef.current) {
            inputRef.current.value = "";
          }
        }}
      >
        + Add
      </StyledButton>
    </TodoInputForm>
  );
};

const mapDispatchToProps = (dispatch: Dispatch<ActionType>) => ({
   addTodo: (text: string) => dispatch({type: ADD_TODO, id: idEnc++, text}),
    dispatch
});
export default connect(null, mapDispatchToProps)(TodoInput);
