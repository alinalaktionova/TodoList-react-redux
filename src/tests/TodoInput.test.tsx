import React from "react";
import { Provider } from "react-redux";
import TodoInput from "../components/TodoInput";
import ReactTestUtils from "react-dom/test-utils";
import configureStore, { MockStore } from "redux-mock-store";
import renderer from "react-test-renderer";
import { addTodo } from "../actions/actionCreators";

const mockStore = configureStore([]);

describe("My Connected React-Redux Component", () => {
  let store: MockStore;
  let component: any;
  beforeEach(() => {
    store = mockStore({ todos: [] });
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store}>
        <TodoInput />
      </Provider>
    );
  });
  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
  it("should dispatch an action on button click", () => {
    renderer.act(() => {
      component.root.findByType("button").props.onClick();
    });

    renderer.act(() => {
      component.root.findByType("input");
      ReactTestUtils.Simulate.change(component);
    });
    ReactTestUtils.Simulate.click(
      component,
      store.dispatch(addTodo(component))
    );
  });
});
