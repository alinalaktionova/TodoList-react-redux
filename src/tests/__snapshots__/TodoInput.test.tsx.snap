// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="sc-bdVaJa hiKACg"
>
  <input
    className="sc-bwzfXH bjjtcQ"
    onChange={[Function]}
    placeholder="add a todo item"
    type="text"
  />
  <button
    className="sc-htpNat hIreTc"
    onClick={[Function]}
    type="button"
  >
    + Add
  </button>
</div>
`;
